pkg/serverless/proxy/transport.go

```diff
@@ -0,0 +1,75 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package proxy
+
+import (
+	"bytes"
+	"errors"
+	"net/http"
+	"net/http/httputil"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type proxyTransport struct {
+	currentInvocationDetails *invocationDetails
+	processor                invocationProcessor
+}
+
+func (p *proxyTransport) RoundTrip(request *http.Request) (*http.Response, error) {
+	log.Debug("[proxy] new request to %s", request.URL)
+
+	// enrich the currentInvocationDetails object
+	requestProcessor(p.currentInvocationDetails, request)
+	if p.currentInvocationDetails.isComplete() {
+		p.processor.process(p.currentInvocationDetails)
+	}
+
+	response, err := http.DefaultTransport.RoundTrip(request)
+	if err != nil {
+		log.Error("could not forward the request", err)
+		return nil, err
+	}
+
+	dumpedResponse, err := httputil.DumpResponse(response, true)
+	if err != nil {
+		log.Error("could not dump the response")
+
+		return nil, err
+	}
+
+	// extract only the payload as headers can be retrieved without inspecting the response
+	indexPayload := bytes.Index(dumpedResponse, []byte("\r\n\r\n"))
+	if indexPayload == -1 {
+		return nil, errors.New("invalid payload format")
+	}
+
+	// enrich the currentInvicationDetails when /next response is received
+	if request.Method == "GET" && strings.HasSuffix(request.URL.String(), "/next") {
+		p.currentInvocationDetails.startTime = time.Now()
+		p.currentInvocationDetails.invokeHeaders = response.Header
+		p.currentInvocationDetails.invokeEventPayload = string(dumpedResponse[indexPayload:])
+	}
+
+	return response, nil
+}
+
+func requestProcessor(invocationDetails *invocationDetails, request *http.Request) {
+	if request.Method == "GET" && strings.HasSuffix(request.URL.String(), "/next") {
+		invocationDetails.reset()
+	} else if request.Method == "POST" && strings.HasSuffix(request.URL.String(), "/response") {
+		invocationDetails.endTime = time.Now()
+		invocationDetails.isError = false
+	} else if request.Method == "POST" && strings.HasSuffix(request.URL.String(), "/error") {
+		invocationDetails.endTime = time.Now()
+		invocationDetails.isError = true
+		// TODO create and send the error enhanced metric here
```

Comment:
Is there any way we can avoid adding business logic related to tracing/metrics to the proxy itself, and instead provide some kind of API to use the proxy? In the Lambda Libraries, we provide hooks at the beginning and end of the invocation. Would it be possible to create a similar API for the proxy?

Reply:
More along these lines -- what is the difference in purpose between the `process` function in `invocation_processor` versus this one in `transport`?

Reply:
tl;dr It would be nice if other engineers working with the proxy never have to touch the proxy module itself to add features

Reply:
yeah that's the goal with the `invocationProcessor` interface.
you just need to implement the `process` method and you're abstracted from the proxy logic here : https://github.com/DataDog/datadog-agent/pull/9988/files#diff-e0b3c1fdd1d581893bfa049c2ae00afe78161af139688f9f558456c20e602bbbR18